LibWeb/URL: Add `strip_trailing_spaces_from_an_opaque_path()`

kemzeb opened the pull request:
Adds https://url.spec.whatwg.org/#potentially-strip-trailing-spaces-from-an-opaque-path.

Tested the code paths in `set_search()` and `set_hash()`  and they appear fine, though I think the design may be questionable, specifically with exposing more methods and delegating the work in those methods to `AK::URL`.

shannonbooth commented on AK/URL.h:
```diff
@@ -84,6 +84,9 @@ class URL {
     size_t path_segment_count() const { return m_paths.size(); }
 
+    // A URL has an opaque path if its path is a URL path segment.
+    bool has_opaque_path() const { return path_segment_count() == 1; }
```
I think it is quite confusing, but I believe `cannot_be_a_base_url` may just be equivalent to "has an opaque path" or at least, that's how we seem to treat it at the moment?). Also, is it correct to say that it has an opaque path if the path segment count is _exactly_ one?

Ref: https://github.com/SerenityOS/serenity/blob/851ac682efb893f206e0dab793d5480ee5473e58/AK/URLParser.cpp#L920

kemzeb replied:
Thanks for looking into this!

>  think it is quite confusing, but I believe cannot_be_a_base_url may just be equivalent to "has an opaque path" or at least, that's how we seem to treat it at the moment?).

Yeah I found calls to this method while reading the `LibWeb::URL` code and wondered if I could use that instead, but was deterred when I saw that anyone could set the value since it also had a setter. I added that method to `AK::URL` since it seemed to align with how the [spec ](https://url.spec.whatwg.org/#url-opaque-path) defined it.

I also feel that based off the name of `cannot_be_a_base_url `, one could see both `https://example.com/a` (which has an opaque path) and `https://example.com/a/b/c` (not an opaque path) as not being base URLs. Let me know what you think!

> Also, is it correct to say that it has an opaque path if the path segment count is exactly one?

The spec phrased it in a way that would mean their could be more than one path fragment (even though that's not the case since that goes against its definition). I was confused by this too, but thought it was safer to be consistent with the spec. I'll make a change for that.

shannonbooth replied:
I found one relevant change to the spec, the old phrasing "cannot be a base URL" in certain places seem to have been renamed: https://github.com/whatwg/url/commit/fbaa03cb19ee5718953f5f6d179e0339e31f1ede

We have artifacts of that change in other locations in our code, for example: https://github.com/SerenityOS/serenity/blob/ec416563a896f188072bac6e86a3234888a8f231/Userland/Libraries/LibWeb/URL/URL.cpp#L282

But if you view: https://url.spec.whatwg.org/#dom-url-hostref-for-dom-url-host%E2%91%A0 you'll notice that the phrasing has been updated per that above commit.

When I get the time a little later, I'll try dig some more to understand. Seems we've become out of date with the spec at some stage, and we need to go through and fix up the spec comments / terminology at some stage to make this more clear.

kemzeb replied:
Do you think I should still include `is_opaque_path()` in `AK::URL` or maybe just have `LibWeb::URL::is_opaque_path()` call on `AK::URL::cannot_be_a_base_url()` for the time being?

shannonbooth replied:
My guess is that we may need a new function for this, but would I need to spend some time reading the spec to properly understand what we should be doing here.

shannonbooth replied:
Okay, I've found some time to look at this properly. I don't think this implementation of `has_opaque_path` is correct. It looks like this was originally changed in: https://github.com/whatwg/url/pull/655

Later on, this was refined in: https://github.com/whatwg/url/pull/736

For now, I think it's okay to use `cannot_be_a_base_url`. This explains some of the issues you were having with: https://github.com/SerenityOS/serenity/pull/21021#discussion_r1321260808

Probably we need to catch up with the spec a little bit for this all to make a little more sense. Trying to figure out at the moment what changes we need to be making.

shannonbooth commented on Userland/Libraries/LibWeb/URL/URL.cpp:
```diff
@@ -535,6 +539,31 @@ bool host_is_domain(AK::URL::Host const& host)
+    if (url.query().has_value())
+        return {};
+
+    // 4. Remove all trailing U+0020 SPACE code points from url’s URL’s path.
```
This spec step states we should be trimming space code units, not ascii whitespace - also, we don't need to return failure here on OOM, we can just `MUST` the result and crash instead if we run out of memory.

shannonbooth commented on Userland/Libraries/LibWeb/URL/URL.h:
```diff
@@ -55,16 +55,23 @@ class URL : public Bindings::PlatformObject {
+
+    ErrorOr<String> path_segment_at_index(size_t index) { return String::from_deprecated_string(m_url.path_segment_at_index(index)); }
+
+    bool has_opaque_path() { return m_url.has_opaque_path(); }
```
```cpp
    bool has_opaque_path() const { return m_url.has_opaque_path(); }
```

For this, and the other helpers, we can const-qualify these member functons.

shannonbooth commented on Userland/Libraries/LibWeb/URL/URL.cpp:
```diff
@@ -450,17 +451,18 @@ WebIDL::ExceptionOr<String> URL::hash() const
 
-        // FIXME: 2. Potentially strip trailing spaces from an opaque path with this.
+        // 2. Potentially strip trailing spaces from an opaque path with this.
+        TRY(strip_trailing_spaces_from_an_opaque_path(*this));
```
Please add a test for this which would previously fail. Potentially a good spot to add a test would be: https://github.com/SerenityOS/serenity/blob/master/Tests/LibWeb/Text/input/URL/url.html

shannonbooth commented:
Interestingly, seems like there is a change being proposed that is removing this section of the spec: https://github.com/whatwg/url/pull/785 (only up for discussion right now)

kemzeb commented:
Nice find! From reading it (and my basic knowledge of character encoding that I should to work on), I believe some of the possible approaches they give would not have *too* much of an impact on my implementation.

shannonbooth commented on Tests/LibWeb/Text/expected/URL/strip_trailing_spaces_from_opaque_path.txt:
```diff
@@ -0,0 +1,4 @@
+pathname => '/well    '
+pathname => '/well'
```
The results for this test don't match the results which I see in chrome or firefox. I believe that this is a bug in our implementation.

A space in the spec is in the percent encoding for "Path", and I believe that we are percent encoding here: https://github.com/SerenityOS/serenity/blob/ec416563a896f188072bac6e86a3234888a8f231/AK/URLParser.cpp#L1539

I think that the issue is that we are percent decoding the path in Web/URL.cpp, which is producing wrong results.

kemzeb replied:
Not sure if this is the problem, but I do see that we percent decode in `AK::serialize_path()`:

https://github.com/SerenityOS/serenity/blob/ec416563a896f188072bac6e86a3234888a8f231/AK/URL.cpp#L270

shannonbooth replied:
Yup, I think that is the problem. Inside of web specs we should probably not be doing the percent decode. Outside of it, we probably should

kemzeb replied:
I'm pretty sure this is the relevant [spec ](https://url.spec.whatwg.org/#url-path-serializer)to `AK::URL::serialize_path()`. There's no explicit mention of a need to percent decode the path during serialization

kemzeb replied:
Continuing my previous [comment](https://github.com/SerenityOS/serenity/pull/21021#issuecomment-1713346790), I also tried to just outright ignore the first step of this method, but that too causes the test suite to hang for some reason.

Oddly enough, reverting the first step back to its old self (i.e. using `cannot_be_a_base_url()`) does not cause any test suite hangs.

kemzeb replied:
One other note; I forgot to account for the URL tests in `AK`. We seem to fail in quite a few: mainly because of the missing slash prefix issue in step 1 and characters being escaped (I tested the failing `unicode` test case specifically on Chrome and Firefox and they escape their pathnames, which is not what we expect in this test case).

If you wish, we could try to tackle these issues within another PR or I could use my workaround of reverting step one of `serialize_path()` back to using `cannot_be_a_base_url()` since this doesn't cause the `LibWeb` test suite to hang.

kemzeb commented:
Recent snapshot implements `AK::serialize_path()` to spec, but not sure if the spec is correct. "If url has an [opaque path](https://url.spec.whatwg.org/#url-opaque-path), then return url’s [path](https://url.spec.whatwg.org/#concept-url-path)." Why don't we prefix a `/`?

Even weirder, when I tried to do it myself by returning a `DeprecatedString::formatted("/{}", m_paths[0])`, the test suite hangs. Not sure what's going on, but the test suite should fail right now since the expected opaque pathnames won't have a `/` prefixed.

kemzeb commented:
The most recent change removes `AK::URL::is_opaque_path()`. It also reverts `AK::URL::serialize_path()` back to using `cannot_be_a_base_url()` as its first step because of the problems mentioned in this [comment](https://github.com/SerenityOS/serenity/pull/21021#issuecomment-1713346790).

I also had to modify 2 `AK/TestURL.cpp` test cases since they fail, but appear to be incorrect. Running the test cases "unicode" and "file_url_with_encoded_characters" on Chrome by doing the following in the DevTools console:
```js
let url = new URL("http://example.com/_ünicöde_téxt_©");
undefined
url.pathname
'/_%C3%BCnic%C3%B6de_t%C3%A9xt_%C2%A9'

let url = new URL("http://example.com/a%23test");
undefined
url.pathname
'/a%23test'
```
their expected output follows inline with the changes I made to these test cases.

I also made changes to the `strip_trailing_spaces_from_an_opaque_path()` do to the the removal of `AK::URL::is_opaque_path()`. I also changed the expected test cases to have escaped space characters.

kemzeb commented:
My recent change fixes a comment typo

shannonbooth commented on Userland/Libraries/LibWeb/URL/URL.cpp:
```diff
@@ -535,6 +537,29 @@ bool host_is_domain(AK::URL::Host const& host)
+{
+    // 1. If url’s URL does not have an opaque path, then return.
+    // FIXME: Reimplement this step once we modernize the URL implementation to the meet spec.
+    if (url.cannot_be_a_base_url())
```
Cannot be a base URL is in my understanding equivalent to an opaque path. So this check seems inverted to me. For reference, the web platform tests for this spec change don't match up with the test which is being added here: https://github.com/web-platform-tests/wpt/pull/37556/files

kemzeb replied:
Thanks for spotting this. From looking at the tests you provided and scouring the specs, I found that [special URL paths](https://url.spec.whatwg.org/#concept-url-path) cannot be an opaque path (see the `Note` section beneath the link).  The changes I made to the JS tests should reflect this.

shannonbooth commented:
> The most recent change removes `AK::URL::is_opaque_path()`. It also reverts `AK::URL::serialize_path()` back to using `cannot_be_a_base_url()` as its first step because of the problems mentioned in this [comment](https://github.com/SerenityOS/serenity/pull/21021#issuecomment-1713346790).
> 
> I also had to modify 2 `AK/TestURL.cpp` test cases since they fail, but appear to be incorrect. Running the test cases "unicode" and "file_url_with_encoded_characters" on Chrome by doing the following in the DevTools console:
> 
> ```js
> let url = new URL("http://example.com/_ünicöde_téxt_©");
> undefined
> url.pathname
> '/_%C3%BCnic%C3%B6de_t%C3%A9xt_%C2%A9'
> 
> let url = new URL("http://example.com/a%23test");
> undefined
> url.pathname
> '/a%23test'
> ```
> 
> their expected output follows inline with the changes I made to these test cases.
> 
> I also made changes to the `strip_trailing_spaces_from_an_opaque_path()` due to the the removal of `AK::URL::is_opaque_path()`. I also changed the expected test cases to have URL-encoded space characters.

I think we need to preserve the behaviour of the existing AK tests here. Existing code in serenity _outside_ of Web specs expect that the path is percent decoded. I removed this flag in: db5ad0c2b0229084f762cd35392cba4f6c3f56bd since nothing was actually setting that flag, and also fixed some places that were percent decoding when we shouldn't have.

But it appears I shouldn't have removed that flag for "serialized_path" (or we need a different API which is a bit more clear?) - we need to be setting it here to follow web specs but _not_ for serenity code outside of LibWeb which expected a percent-decoded path.

kemzeb commented:
I should note that I changed the signature of `LibWeb::URL::path_segment_at_index()` to just directly return a `DeprecatedString`. I also added `LibWeb::URL::set_paths()` which delegates its work to `AK::URL::set_paths()` because `LibWeb::URL::set_pathname()` returns immediately if it finds a URL that is an opaque path.

shannonbooth reviewed (approved):
Looks reasonable to me

ADKaster commented on AK/URL.cpp:
```diff
@@ -260,16 +260,25 @@ bool URL::is_special_scheme(StringView scheme)
-        builder.append('/');
-        builder.append(percent_decode(path));
+
+    // Let output be the empty string.
```
Hm. Kind of weird that the spec has a bulleted list for this instead of a numbered list. Might be worth a spec issue in https://github.com/whatwg/html  if you're feeling speccy :)

kemzeb replied:
Yea I wasn't too sure if this was intentional too. I'll file an issue there 👍

kemzeb replied:
Opened an issue at https://github.com/whatwg/url/issues/786
